File: examples/app/graceful-server.go
// graceful-server.go — Production Go HTTP server for use with dpivot.
//
// Demonstrates:
//   1. /health/live  — liveness probe (process alive)
//   2. /health/ready — readiness probe (DB + cache reachable)
//   3. SIGTERM drain — completes in-flight requests before exiting
//   4. Idempotent handler pattern via Idempotency-Key header
//   5. Structured JSON logging
//
// Build: go build -o server ./examples/app/graceful-server.go
// Run:   DATABASE_URL=postgres://... REDIS_URL=redis://... ./server

package main

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"
)

// ── Config ────────────────────────────────────────────────────────────────────

var (
	port          = envOr("PORT", "3000")
	drainTimeout  = envDuration("DRAIN_TIMEOUT", 30*time.Second)
)

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func envDuration(key string, def time.Duration) time.Duration {
	if v := os.Getenv(key); v != "" {
		d, err := time.ParseDuration(v)
		if err == nil {
			return d
		}
	}
	return def
}

// ── Structured logger ─────────────────────────────────────────────────────────

func logJSON(level, event string, fields map[string]interface{}) {
	fields["level"] = level
	fields["event"] = event
	fields["ts"] = time.Now().UTC().Format(time.RFC3339Nano)
	data, _ := json.Marshal(fields)
	fmt.Fprintf(os.Stderr, "%s\n", data)
}

// ── Idempotency cache (in-memory; replace with Redis in production) ───────────

type idempotencyStore struct {
	// In production use Redis: SET idem:{key} {result} EX 86400 NX
	// This in-memory stub demonstrates the pattern.
	m map[string]idempotentResult
}

type idempotentResult struct {
	Status int
	Body   []byte
}

var idemStore = &idempotencyStore{m: make(map[string]idempotentResult)}

func (s *idempotencyStore) Get(key string) (idempotentResult, bool) {
	v, ok := s.m[key]
	return v, ok
}

func (s *idempotencyStore) Set(key string, r idempotentResult) {
	s.m[key] = r
}

// ── Health checks ─────────────────────────────────────────────────────────────

// checkDeps verifies that the application's dependencies (DB, cache) are
// reachable. Called by /health/ready. Returns nil when all deps are healthy.
func checkDeps(ctx context.Context) error {
	// TCP dial to postgres (replace with actual DB ping in production).
	dbURL := os.Getenv("DATABASE_URL")
	if dbURL != "" {
		dialer := &net.Dialer{}
		conn, err := dialer.DialContext(ctx, "tcp", dbURL)
		if err != nil {
			return fmt.Errorf("db unreachable: %w", err)
		}
		conn.Close()
	}

	// TCP dial to Redis.
	redisURL := os.Getenv("REDIS_URL")
	if redisURL != "" {
		dialer := &net.Dialer{}
		conn, err := dialer.DialContext(ctx, "tcp", redisURL)
		if err != nil {
			return fmt.Errorf("redis unreachable: %w", err)
		}
		conn.Close()
	}

	return nil
}

// ── Handlers ──────────────────────────────────────────────────────────────────

func handleLive(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]interface{}{
		"status": "ok",
		"pid":    os.Getpid(),
	})
}

func handleReady(w http.ResponseWriter, r *http.Request) {
	ctx, cancel := context.WithTimeout(r.Context(), 3*time.Second)
	defer cancel()

	if err := checkDeps(ctx); err != nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]interface{}{
			"status": "not_ready",
			"error":  err.Error(),
		})
		return
	}
	writeJSON(w, http.StatusOK, map[string]interface{}{"status": "ready"})
}

// handleCreateOrder demonstrates an idempotent POST handler.
// Clients must send Idempotency-Key: <uuid> with every request.
// Duplicate requests return the same response without re-executing the mutation.
func handleCreateOrder(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	key := r.Header.Get("Idempotency-Key")
	if key == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{
			"error": "Idempotency-Key header is required",
		})
		return
	}

	// Return cached response for duplicate requests.
	if cached, ok := idemStore.Get(key); ok {
		w.Header().Set("X-Idempotent-Replay", "true")
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(cached.Status)
		w.Write(cached.Body) //nolint:errcheck
		return
	}

	// Process exactly once — decode and persist the result.
	var order map[string]interface{}
	if err := json.NewDecoder(r.Body).Decode(&order); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid JSON"})
		return
	}

	id := newID()
	result, _ := json.Marshal(map[string]string{"id": id, "status": "created"})
	idemStore.Set(key, idempotentResult{Status: http.StatusCreated, Body: result})

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	w.Write(result) //nolint:errcheck
}

func newID() string {
	b := make([]byte, 8)
	rand.Read(b) //nolint:errcheck
	return hex.EncodeToString(b)
}

func writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v) //nolint:errcheck
}

// ── Main ──────────────────────────────────────────────────────────────────────

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/health/live", handleLive)
	mux.HandleFunc("/health/ready", handleReady)
	mux.HandleFunc("/orders", handleCreateOrder)

	srv := &http.Server{
		Addr:    ":" + port,
		Handler: mux,

		// Timeouts prevent slow-client attacks and resource exhaustion.
		ReadHeaderTimeout: 5 * time.Second,
		ReadTimeout:       30 * time.Second,
		WriteTimeout:      30 * time.Second,
		IdleTimeout:       120 * time.Second,
	}

	// Start in background.
	go func() {
		logJSON("info", "server_start", map[string]interface{}{"port": port})
		if err := srv.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
			logJSON("error", "server_fatal", map[string]interface{}{"err": err.Error()})
			os.Exit(1)
		}
	}()

	// ── Graceful shutdown on SIGTERM / SIGINT ─────────────────────────────────
	// Docker sends SIGTERM when stopping a container.
	// dpivot drains the proxy-side connection first (drain period), then
	// Docker stops the container. This server-side drain ensures in-flight
	// HTTP requests complete even if the proxy drain is shorter than expected.
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGTERM, syscall.SIGINT)
	sig := <-quit

	logJSON("info", "shutdown_start", map[string]interface{}{
		"signal":     sig.String(),
		"drain_secs": drainTimeout.Seconds(),
	})

	ctx, cancel := context.WithTimeout(context.Background(), drainTimeout)
	defer cancel()

	// Shutdown stops the listener and waits for active handlers to complete.
	if err := srv.Shutdown(ctx); err != nil {
		logJSON("warn", "shutdown_timeout", map[string]interface{}{"err": err.Error()})
		os.Exit(1)
	}

	logJSON("info", "shutdown_complete", map[string]interface{}{})
}

File: examples/app/graceful-server.js
/**
 * Node.js production server with graceful SIGTERM handling for dpivot.
 *
 * Key behaviours implemented here:
 *   1. /health/live  — liveness probe (process running)
 *   2. /health/ready — readiness probe (DB + Redis reachable)
 *   3. SIGTERM drain — stop accepting, finish in-flight requests, then exit
 *   4. Idempotent POST pattern via Idempotency-Key header
 *   5. Client-side retry with exponential backoff (see retryFetch below)
 *
 * How it integrates with dpivot:
 *   - Docker HEALTHCHECK uses /health/ready
 *   - dpivot rollout waits for /health/ready to return 200
 *   - On SIGTERM (container stop), the server drains for up to 30s
 *   - In-flight requests complete; no 502 errors at the nginx/dpivot layer
 */

'use strict';

const http = require('http');
const { Pool } = require('pg');       // npm i pg
const { createClient } = require('redis'); // npm i redis

// ── Dependencies ──────────────────────────────────────────────────────────────

const db = new Pool({ connectionString: process.env.DATABASE_URL });
const redisClient = createClient({ url: process.env.REDIS_URL });

async function connectDeps() {
  await redisClient.connect();
  await db.query('SELECT 1'); // verify DB is reachable on startup
}

// ── Idempotency store ─────────────────────────────────────────────────────────
// Stores processed Idempotency-Key → response so duplicate requests return
// the same result without re-executing the mutation.

async function getIdempotentResult(key) {
  const cached = await redisClient.get(`idem:${key}`);
  return cached ? JSON.parse(cached) : null;
}

async function saveIdempotentResult(key, result, ttlSeconds = 86400) {
  await redisClient.setEx(`idem:${key}`, ttlSeconds, JSON.stringify(result));
}

// ── Request handler ───────────────────────────────────────────────────────────

async function handleRequest(req, res) {
  const url = new URL(req.url, `http://${req.headers.host}`);

  // Liveness: always 200 while the process is alive.
  if (url.pathname === '/health/live') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ status: 'ok', pid: process.pid }));
    return;
  }

  // Readiness: only 200 when dependencies are healthy.
  if (url.pathname === '/health/ready') {
    try {
      await db.query('SELECT 1');
      await redisClient.ping();
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ status: 'ready' }));
    } catch (err) {
      res.writeHead(503, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ status: 'not_ready', error: err.message }));
    }
    return;
  }

  // Example idempotent POST: POST /orders
  if (url.pathname === '/orders' && req.method === 'POST') {
    const idempotencyKey = req.headers['idempotency-key'];
    if (!idempotencyKey) {
      res.writeHead(400, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'Idempotency-Key header required for POST /orders' }));
      return;
    }

    // Return cached result for duplicate requests — safe during retries.
    const cached = await getIdempotentResult(idempotencyKey);
    if (cached) {
      res.writeHead(cached.status, { 'Content-Type': 'application/json', 'X-Idempotent-Replay': 'true' });
      res.end(JSON.stringify(cached.body));
      return;
    }

    // Process the mutation exactly once.
    const body = await readBody(req);
    const order = JSON.parse(body);
    const result = await db.query(
      'INSERT INTO orders (data, created_at) VALUES ($1, NOW()) RETURNING id',
      [JSON.stringify(order)]
    );
    const response = { id: result.rows[0].id, status: 'created' };

    // Persist result so retries get the same response.
    await saveIdempotentResult(idempotencyKey, { status: 201, body: response });

    res.writeHead(201, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(response));
    return;
  }

  res.writeHead(404);
  res.end();
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on('data', c => chunks.push(c));
    req.on('end', () => resolve(Buffer.concat(chunks).toString()));
    req.on('error', reject);
  });
}

// ── Server lifecycle ──────────────────────────────────────────────────────────

const server = http.createServer((req, res) => {
  handleRequest(req, res).catch(err => {
    console.error({ event: 'request_error', path: req.url, err: err.message });
    if (!res.headersSent) {
      res.writeHead(500, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'internal server error' }));
    }
  });
});

const PORT = parseInt(process.env.PORT || '3000', 10);
const DRAIN_TIMEOUT_MS = parseInt(process.env.DRAIN_TIMEOUT_MS || '30000', 10);

async function start() {
  await connectDeps();
  server.listen(PORT, () => {
    console.log(JSON.stringify({ event: 'server_start', port: PORT }));
  });
}

// ── Graceful shutdown (SIGTERM) ───────────────────────────────────────────────
// Docker / dpivot sends SIGTERM before killing the container.
// Steps:
//   1. Stop accepting new TCP connections (server.close).
//   2. Wait for all in-flight requests to complete (or force-exit on timeout).
//   3. Close DB pool and Redis connection cleanly.

function shutdown(signal) {
  console.log(JSON.stringify({ event: 'shutdown_start', signal, drain_ms: DRAIN_TIMEOUT_MS }));

  // 1. Stop accepting new connections. In-flight requests still run.
  server.close(async () => {
    console.log(JSON.stringify({ event: 'shutdown_complete' }));
    await db.end().catch(() => {});
    await redisClient.quit().catch(() => {});
    process.exit(0);
  });

  // 2. Force exit if drain takes too long (prevents hung containers).
  setTimeout(() => {
    console.error(JSON.stringify({ event: 'shutdown_timeout', drain_ms: DRAIN_TIMEOUT_MS }));
    process.exit(1);
  }, DRAIN_TIMEOUT_MS).unref();
}

process.on('SIGTERM', () => shutdown('SIGTERM'));
process.on('SIGINT',  () => shutdown('SIGINT'));

// Crash loudly on unhandled rejections so the container restarts.
process.on('unhandledRejection', (reason) => {
  console.error(JSON.stringify({ event: 'unhandled_rejection', reason: String(reason) }));
  process.exit(1);
});

start().catch(err => {
  console.error(JSON.stringify({ event: 'startup_failed', err: err.message }));
  process.exit(1);
});

// ── Client-side retry with exponential backoff ────────────────────────────────
// Use this pattern in any service that calls this API.
// Retries are safe for GET and for POST requests that include Idempotency-Key.

async function retryFetch(url, options = {}, maxAttempts = 4) {
  const { idempotencyKey, ...fetchOptions } = options;
  if (idempotencyKey) {
    fetchOptions.headers = {
      ...fetchOptions.headers,
      'Idempotency-Key': idempotencyKey,
    };
  }

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      const res = await fetch(url, fetchOptions);
      // Retry on 502/503/504 (proxy/upstream errors) and 429 (rate limit).
      if ([429, 502, 503, 504].includes(res.status) && attempt < maxAttempts) {
        const backoffMs = Math.min(100 * 2 ** (attempt - 1), 5000) + Math.random() * 100;
        console.warn(`retryFetch: attempt ${attempt} got ${res.status}, retrying in ${backoffMs.toFixed(0)}ms`);
        await new Promise(r => setTimeout(r, backoffMs));
        continue;
      }
      return res;
    } catch (err) {
      if (attempt === maxAttempts) throw err;
      const backoffMs = Math.min(100 * 2 ** (attempt - 1), 5000);
      await new Promise(r => setTimeout(r, backoffMs));
    }
  }
}

module.exports = { retryFetch };
